### PR TITLE
show-achievements-badges-on-members-site

### PR DESCRIPTION
**Pull request by shmbajaj**
### What is the change?

- Create badge, Assign badges, Unassign badges and Get-Assigned badgeIds.
- Superuser can only create, assign and unassign badge.

### Before this change?

- Superuser cannot create, assign and unassign badge.
- `/badges/` used to return empty array of badges.
- `/badges/:username` used to return empty array of assigned badges(not badgeIds). 

### Endpoints

- Get all available badges in DB
```javascript
GET /badges
returns
{
id: firebaseDocumentId <string>;
name: badgeName <string>;
description: badgeDescription <string>;
imageUrl: badgeImageHostedInCloudinary <string>;
createdAt: {date: string; time: string;} <timestamp>;
createdBy: username <string>;
}[]
```
- Get assigned badge to the user
```javascript
GET /user/:id/badges
returns
{
message: successMessage <string>;
badges: assignedBadges Array<Object<Badge>>;
}
```
- Create badge
```javascript
POST /badges
payload
{
name: badgeName <string>;
description: badgeDescription <string>;
file: imageFile[name=badge] <file>;
createdBy: username <string>;
}

returns
{
message: successMessage <string>;
badge: badgeObject Object<Badge>;
}
```
- Assign badges
```javascript
POST /badges/assign
payload
{
userId: user-id <string>;
badgeIds: badgeIdsToBeAssigned Array<string>;
}

returns
{
message: successMessage <string>;
}
```
- UnAssign badges
```javascript
DELETE /badges/unassign
payload
{
userId: user-id <string>;
badgeIds: badgeIdsToBeUnAssigned Array<string>;
}

returns
{
message: successMessage <string>;
}
```

> Close [469](https://github.com/Real-Dev-Squad/website-members/issues/469)
> Website-Api-Contracts: [Website-Api-Contracts](https://github.com/Real-Dev-Squad/website-api-contracts/pull/103)

### \*Dev Tested?

- [X] Yes
- [ ] No

#### Platforms

- [X] Postman

### Frontend Finished

- [ ] Yes
- [X] No
<small>**Yet To Start**</small>

### Info
- **Failing tests and code-scanning results has nothing to do with this PR.**

### CODE REVIEW

File: controllers/badges.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      responseMsg = badgeIds.length !== 0 ? "User badges returned successfully!" : "This user does not have any badges";
+      return res.json({ message: responseMsg, badgeIds });
+    } else {
+      return res.boom.notFound("The user does not exist");
+    }
+  } catch (error) {
+    logger.error(`Error while fetching all user badges: ${error}`);
+    return res.boom.serverUnavailable("Something went wrong please contact admin");
+  }
+};
```
Comment (vinayak-trivedi) [inline controllers/badges.js:38]: why do you need only badge id?

File: models/badges.js [hunk 2]
```diff
@@ -27,36 +42,99 @@ const fetchBadges = async ({ size = 100, page = 0 }) => {
...
-
-      snapshot.forEach((item) => {
-        if (item.data()?.users?.includes(userID)) {
-          const { title, description } = item.data();
-          userBadges.push({ title, description });
-        }
-      });
+      const userId = result.user.id;
+      const snapshot = await userBadgeModel.where("userId", "==", userId).get();
+      badgeIds = snapshot.docs.map((doc) => doc.get("badgeId"));
```
Comment (vinayak-trivedi) [inline models/badges.js:57]: why are we only taking badgeId?

File: controllers/badges.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+const getUserBadgeIds = async (req, res) => {
+  try {
+    const { userExists, badgeIds } = await badgeQuery.fetchUserBadgeIds(req.params.username);
+    let responseMsg = "";
+    if (userExists) {
+      responseMsg = badgeIds.length !== 0 ? "User badges returned successfully!" : "This user does not have any badges";
+      return res.json({ message: responseMsg, badgeIds });
+    } else {
+      return res.boom.notFound("The user does not exist");
+    }
```
Comment (vinayak-trivedi) [inline controllers/badges.js:33]: we know that badgeIds is an array, so we can return it directly with the message badges returned successfully, and in the frontend we can check if the array is empty and change the message accordingly, also the following would be much cleaner

c
```suggestion
    const { userExists, badgeIds } = await badgeQuery.fetchUserBadgeIds(req.params.username);

    if (!userExists) {
       return res.boom.notFound("The user does not exist");
    }
      
      return res.json({ message: 'badges returned successfuly, badgeIds });
    
```

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]
Thanks for the PR @shmbajaj, suggested some changes please have a look

File: controllers/badges.js (hunk 3 above)
Comment (shmbajaj) [reply controllers/badges.js:33]: I understood your point, and have make changes accordingly.

File: controllers/badges.js (hunk 1 above)
Comment (shmbajaj) [reply controllers/badges.js:38]: My reasoning was to return badgeIds and map these badgeIds with 
response of `GET /badges` request i.e. badges.

Now I think it's better to return badges only for whom it is asked, 
as the ratio of visit of super-user: normal-user will be low compared to the other.

File: models/badges.js (hunk 2 above)
Comment (shmbajaj) [reply models/badges.js:57]: Refer above comment for this reasoning.

File: controllers/badges.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+ * @returns {message} - badges unassigned
+ */
+async function deleteUserBadges(req, res) {
+  try {
+    const { username } = req.params;
+    const { badgeIds } = req.body;
+    const result = await fetchUser({ username });
+    if (!result.userExists) {
+      throw new Error("Failed to assign badges, user does not exsit");
+    }
```
Comment (vinayak-trivedi) [inline controllers/badges.js:102]: should this be `failed to delete badge, the user doesn't exist`

**Review by vinayak-trivedi — DISMISSED** [review]
LGTM

**Comment by heyrandhir** [comment]
Tests are failing. can you please look into it @shmbajaj . Have the data models and api contracts also been updated ? Please update it if not.

File: models/badges.js [hunk 5]
```diff
@@ -27,36 +27,106 @@ const fetchBadges = async ({ size = 100, page = 0 }) => {
...
+    const docIds = [];
+    const batch = firestore.batch();
+    badgeIds.forEach((badgeId) => {
+      const ref = userBadgeModel.doc();
+      const id = ref.id;
+      batch.create(ref, { userId, badgeId });
+      docIds.push(id);
+    });
+    await batch.commit();
+    return { docIds };
```
Comment (heyrandhir) [inline models/badges.js:97]: The batch operation has a limit of 500 documents. we should add it in logger as well if we find more than 500 docs are there to update.

File: models/badges.js [hunk 6]
```diff
@@ -27,36 +27,106 @@ const fetchBadges = async ({ size = 100, page = 0 }) => {
...
+  try {
+    const docIds = [];
+    const snapshot = await userBadgeModel.where("userId", "==", userId).where("badgeId", "in", badgeIds).get();
+    const batch = firestore.batch();
+    snapshot.forEach((doc) => {
+      docIds.push(doc.id);
+      batch.delete(doc.ref);
+    });
+    await batch.commit();
+    return { docIds };
```
Comment (heyrandhir) [inline models/badges.js:119]: The batch operation has a limit of 500 documents. we should add it in logger as well if we find more than 500 docs are there to update.

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: models/badges.js (hunk 6 above)
Comment (shmbajaj) [reply models/badges.js]: Thanks, for pointing this out. Using functions `chunks` and `assignUnassignBadgesInBulk` writes will happen in chunks of 500 documents per batch.

`example:`
```js
 document-size is 2000
 chunks-size is 500
 totoal-unqiue-batches will be 4
 ```
 
 [Function: assignUnassignBadgesInBulk](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/d935f6802a95e7f2d9ee71cd84957e0c0548d0af)
 [Function: chunks](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/e8dcb542649a5cabbb8392c523cd5f9521ef6252)

File: models/badges.js (hunk 5 above)
Comment (shmbajaj) [reply models/badges.js]: Reffer below comment: https://github.com/Real-Dev-Squad/website-backend/pull/832#discussion_r1070247617

File: controllers/badges.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    const { badgeIds } = req.body;
+    const result = await fetchUser({ username });
+    if (!result.userExists) {
+      throw new Error("Failed to delete badges, user does not exsit");
+    }
+    const userId = result.user.id;
+    await badgeQuery.unAssignBadges({ userId, badgeIds });
+    return res.json({
+      message: "Badges un-assigned successfully.",
+    });
```
[the remark this answers is not in the document]
Comment (shmbajaj) [reply controllers/badges.js:112]: I have handled this scenario please check here: https://github.com/Real-Dev-Squad/website-backend/pull/832/files#diff-e347a36f189bf492829f9a55579ad545f41ecbe17a8d82e64814890eb8d2a72eR113

**Review by heyrandhir — APPROVED** [review]
LGTM. Great handling of batch updates. Good Work @shmbajaj with the documentation as well.

File: controllers/badges.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+ * Unassign badges
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ * @returns {Object}: <message: string> - badges unassigned
+ */
+async function deleteUserBadges(req, res) {
+  try {
+    const { username } = req.params;
+    const { badgeIds } = req.body;
+    const userId = await getUserId(username);
```
Comment (ivinayakg) [inline controllers/badges.js:101]: same as above `/badges/user/id`

File: routes/badges.js [hunk 9]
```diff
@@ -1,8 +1,36 @@
...
-const badge = require("../controllers/badge.js");
+const authenticate = require("../middlewares/authenticate");
+const badgeValidator = require("../middlewares/validators/badges");
+const { upload } = require("../utils/multer");
+const badge = require("../controllers/badges");
+const authorizeRoles = require("../middlewares/authorizeRoles");
+const { SUPERUSER } = require("../constants/roles");
 
 router.get("/", badge.getBadges);
 router.get("/:username", badge.getUserBadges);
```
Comment (ivinayakg) [inline routes/badges.js:11]: `/user/:username`

File: services/imageService.js [hunk 10]
```diff
@@ -32,6 +32,30 @@ const uploadProfilePicture = async ({ file, userId, coordinates }) => {
   }
 };
 
+/**
+ * upload badge image to cloudinary
+ * @param file { Object }: File object
+ * @param badgeName { string }: Badge name
+ * @return image { Object }: id and imageUrl
+ */
+async function uploadBadgeImage({ file, badgeName }) {
```
Comment (ivinayakg) [inline services/imageService.js:41]: create an issue, to make this generic

File: constants/badges.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+const SUCCESS_MESSAGES = {
+  controllers: {
+    deleteUserBadges: "Badges un-assigned successfully",
+    postUserBadges: "Badges assigned successfully",
+    postBadge: "Badge created successfully",
+    getUserBadges: "User Badges returned succesfully",
+    getBadges: "Badges returned successfully",
+  },
+};
```
Comment (vinayak-trivedi) [inline constants/badges.js:41]: all the constants should be capitalized

File: services/imageService.js (hunk 10 above)
Comment (shmbajaj) [reply services/imageService.js:41]: okay, it will be done!!

File: routes/badges.js (hunk 9 above)
Comment (shmbajaj) [reply routes/badges.js]: Route: [GET /user/:id/badges](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/3f4b91b4ab48a34d81047be1aab54b62db5a9b01)

File: constants/badges.js (hunk 11 above)
Comment (shmbajaj) [reply constants/badges.js]: Transform To Uppercase: [All constansts](https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/dbcad66970bced2124ea66f5cd84ecee0df201e3)

File: controllers/badges.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+ * @param res {Object} - Express response object
+ * @returns {Object}: <message: string> - badges assigned
+ */
+// INFO: badgeIds are not validated, hence user can be assigned same badge multiple times
+// TODO: add check for isBadgeIdExsist
+async function postUserBadges(req, res) {
+  try {
+    const { username } = req.params;
+    const { badgeIds } = req.body;
+    const userId = await getUserId(username);
```
[the remark this answers is not in the document]
Comment (shmbajaj) [reply controllers/badges.js]: https://github.com/Real-Dev-Squad/website-backend/pull/832/commits/d6b60276bc8d7d618495ad33955d48152843f69d

File: controllers/badges.js (hunk 8 above)
Comment (shmbajaj) [reply controllers/badges.js]: PLease refer above comment!!

File: models/badges.js [hunk 13]
```diff
@@ -1,62 +1,132 @@
...
+    }
+    const badgeDocReferences = badgeIdsSnapshot.docs.map((doc) => {
+      const badgeId = doc.get("badgeId");
+      return firestore.doc(`badges/${badgeId}`);
+    });
+    // INFO: getAll accepts unpacked array
+    // TODO: check getAll limitiations
+    const badgesSnapshot = await firestore.getAll(...badgeDocReferences);
+    const badges = badgesSnapshot.map((doc) => convertFirebaseDocumentToBadgeDocument(doc.id, doc.data()));
+    return { badges };
```
Comment (ivinayakg) [inline models/badges.js:54]: why return an object, can't we just return the output array as it is, or is there anything else?

**Review by ivinayakg — DISMISSED** [review]

File: models/badges.js [hunk 14]
```diff
@@ -1,62 +1,132 @@
...
+}
+
+/**
+ * Add badge to firestore
+ * @param  badgeInfo { Object }: has badge name, description, imageUrl and createdBy
+ * @return {Promise}: <{id: string, createdAt: {date: string, time: string}, data: any> returns badge object
  */
+async function createBadge(badgeInfo) {
+  try {
+    const createdAt = admin.firestore.Timestamp.now();
```
[the remark this answers is not in the document]
Comment (shmbajaj) [reply models/badges.js:68]: I cannot recall exact reason for not using `function: Date()`, 
but I guess both format had a difference so, I choose to go with `Firebase.Timestamp.now()`.

File: models/badges.js (hunk 13 above)
Comment (shmbajaj) [reply models/badges.js:54]: It is left by mistake, previously I was sending more info but later 
decided not to send as that will be unneccessary info. 
I will remove in next iteration or when adding rate-limiting middelware.

File: controllers/badges.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+/**
+ * Assign badges
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ * @returns {Object}: <message: string> - badges assigned
+ */
+// INFO: badgeIds are not validated, hence user can be assigned same badge multiple times
+// TODO: add check for isBadgeIdExsist
+async function postUserBadges(req, res) {
```
[the remark this answers is not in the document]
Comment (shmbajaj) [reply controllers/badges.js:76]: Name can be `assignUserBadges`, but I followed exisiting code for controllers name. 
So, used `get-name` or `post-name`.

File: test/integration/badges.test.js [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+    userId = await addUser(superUser);
+    jwt = authService.generateAuthToken({ userId });
+  });
+
+  after(async function () {
+    await cleanDb();
+  });
+
+  describe("GET /badges", function () {
+    // TODO: write a more strong test
```
Comment (heyrandhir) [inline test/integration/badges.test.js:41]: is this something pending ?

File: test/integration/badges.test.js [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+          expect(response).to.have.status(400);
+          expect(response.body.message).to.contains(ERROR_MESSAGES_VALIDATORS.CREATE_BADGE.FILE_IS_MISSING);
+          expect(response.body.error).to.equal("Bad Request");
+
+          return done();
+        });
+    });
+
+    it("Should return API payload failed validation, createdBy is required", function (done) {
+      // TODO: write a more strong test
```
Comment (heyrandhir) [inline test/integration/badges.test.js:98]: are we missing anything here as well

File: test/integration/badges.test.js [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+            return done();
+          }
+          expect(response).to.have.status(401);
+          expect(response.body.message).to.contains(MISC.UNAUTHENTICATED_USER);
+          expect(response.body.error).to.equal("Unauthorized");
+          return done();
+        });
+    });
+
+    it("Should return file is missing", function (done) {
```
Comment (heyrandhir) [inline test/integration/badges.test.js:80]:
```suggestion
    it("Should return an error if file missing", function (done) {

**Review by heyrandhir — CHANGES_REQUESTED** [review]

**Review by heyrandhir — APPROVED** [review]

File: controllers/badges.js [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+const getBadges = async (req, res) => {
+  try {
+    const allBadges = await badgeQuery.fetchBadges(req.query);
+    return res.json({
+      message: CONTROLLERS_SUCCESS_MESSAGES.GET_BADGES,
+      badges: allBadges,
+    });
+  } catch (error) {
+    logger.error(`${CONTROLLERS_ERROR_MESSAGES.GET_BADGES}: ${error}`);
+    return res.boom.badRequest(CONTROLLERS_ERROR_MESSAGES.GET_BADGES);
```
Comment (satyam73) [inline controllers/badges.js:23]: You can also concatenate error message here

File: middlewares/validators/badges.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+      userId: joi.string().required(),
+      badgeIds: joi.array().min(1).items(joi.string().required()).unique().required(),
+    });
+  try {
+    const { badgeIds, userId } = req.body;
+    await schema.validateAsync({ userId, badgeIds });
+    next();
+  } catch (error) {
+    logger.error(`${ASSIGN_OR_REMOVE_BADGES.VALIDATON_FAILED}: ${error}`);
+    res.boom.badRequest(`${API_PAYLOAD_VALIDATION_FAILED}, ${error.details?.[0]?.message}`);
```
Comment (satyam73) [inline middlewares/validators/badges.js:55]: null coalesing can be used as above

**Review by Muralidhar22 — COMMENTED** [review]
Tests are failing because of validation of query params of `users` api

**Comment by shmbajaj** [comment]
> Tests are failing because of validation of query params of `users` api(`search` param is not included)

@Muralidhar22 have you raised any ticket, for this issue.
If not then we can raise another PR for this, and also what you have mentioned is not related to this PR.

**Pull request merged**